### PR TITLE
heap PR jemalloc analysis

### PR DESCRIPTION
**Pull request by soez**

### CODE REVIEW

File: libr/core/linux_heap_jemalloc.c [hunk 1]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
 }
 
+static void GH(jemalloc_get_runs)(RCore *core, const char *input) {
+	switch (input[0]) {
+	case ' ':
+		{
+			int i, j;
+			GHT addr = UT64_MAX, size = UT64_MAX, mapbits = UT64_MAX, flags = UT64_MAX, mask = UT64_MAX, chunk = UT64_MAX, misc_offset=UT64_MAX;
+			GHT map = UT64_MAX, map_misc_offset = UT64_MAX, chunksize_mask = UT64_MAX, npages = UT64_MAX, map_bias = UT64_MAX;
+			arena_chunk_t *c = R_NEW0 (arena_chunk_t);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: check for null

File: libr/core/linux_heap_jemalloc.c [hunk 2]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+
+			input += 1;
+			chunk = strstr (input, "0x") ? (GHT)strtol (input, NULL, 0) : (GHT)strtol (input, NULL, 16);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: ut64 chunk = r_num_math (core->num, input);

File: libr/core/linux_heap_jemalloc.c [hunk 3]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+
+			if (GH(r_resolve_jemalloc)(core, "je_chunk_npages", &npages)) {
+				r_core_read_at (core, npages, (ut8*)&npages, sizeof (GHT));
+			} else {
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: handle error in the first branch, and put the core_read_at below outside the else block, return blocks doesnt need an else

File: libr/core/linux_heap_jemalloc.c [hunk 4]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+				eprintf ("Error resolving je_chunk_npages\n");
+				return;
+			}
+
+			if (GH(r_resolve_jemalloc)(core, "je_chunksize_mask", &chunksize_mask)) {
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: same here

File: libr/core/linux_heap_jemalloc.c [hunk 5]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+				eprintf ("Error resolving je_map_bias");
+			}
+
+			if (GH(r_resolve_jemalloc)(core, "je_map_misc_offset", &map_misc_offset)) {
+				r_core_read_at (core, map_misc_offset, (ut8*)&map_misc_offset, sizeof (GHT));
+			} else {
+				eprintf ("Error resolving je_map_misc_offset");
+			}
+
+			printf("map_misc_offset %x\n", map_misc_offset);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: %x numbers must be prefixed with 0x,
missing space before (
dont use printf, use eprintf

File: libr/core/linux_heap_jemalloc.c [hunk 6]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+			mapbits = *(GHT *)&c->map_bits;
+			eprintf ("map_bits: 0x%"PFMTx"\n", mapbits);
+
+			uint32_t offset = r_offsetof(arena_chunk_t, map_bits);
+
+			arena_chunk_map_bits_t *dwords = (size_t*)calloc(sizeof(arena_chunk_map_bits_t), npages);
+			r_core_read_at (core, chunk+offset, (ut8*)dwords, sizeof(arena_chunk_map_bits_t) * npages);
+			printf("map_bits @ 0x%"PFMTx"\n", chunk+offset);
+
+			arena_run_t *r= R_NEW0 (arena_run_t);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: check if allocation fails

File: libr/core/linux_heap_jemalloc.c [hunk 7]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+
+			r_core_read_at (core, chunk, (ut8 *)c, sizeof (arena_chunk_t));
+			mapbits = *(GHT *)&c->map_bits;
+			eprintf ("map_bits: 0x%"PFMTx"\n", mapbits);
+
+			uint32_t offset = r_offsetof(arena_chunk_t, map_bits);
+
+			arena_chunk_map_bits_t *dwords = (size_t*)calloc(sizeof(arena_chunk_map_bits_t), npages);
+			r_core_read_at (core, chunk+offset, (ut8*)dwords, sizeof(arena_chunk_map_bits_t) * npages);
+			printf("map_bits @ 0x%"PFMTx"\n", chunk+offset);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: use eprintf or r_cons_printf

File: libr/core/linux_heap_jemalloc.c [hunk 8]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+				eprintf ("Error resolving je_map_misc_offset");
+			}
+
+			printf("map_misc_offset %x\n", map_misc_offset);
+
+			r_core_read_at (core, chunk, (ut8 *)c, sizeof (arena_chunk_t));
+			mapbits = *(GHT *)&c->map_bits;
+			eprintf ("map_bits: 0x%"PFMTx"\n", mapbits);
+
+			uint32_t offset = r_offsetof(arena_chunk_t, map_bits);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: missing space before (

File: libr/core/linux_heap_jemalloc.c [hunk 9]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+
+			arena_chunk_map_bits_t *dwords = (size_t*)calloc(sizeof(arena_chunk_map_bits_t), npages);
+			r_core_read_at (core, chunk+offset, (ut8*)dwords, sizeof(arena_chunk_map_bits_t) * npages);
+			printf("map_bits @ 0x%"PFMTx"\n", chunk+offset);
+
+			arena_run_t *r= R_NEW0 (arena_run_t);
+
+			for (int pageind = map_bias; pageind < npages; pageind++) {
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: for (int is not c99, please put the int pageind; outside the for

File: libr/core/linux_heap_jemalloc.c [hunk 10]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+				arena_chunk_map_bits_t mapelm = dwords[pageind-map_bias];
+
+				if (mapelm.bits & CHUNK_MAP_ALLOCATED) {
+					uintptr_t elm =  ((arena_chunk_map_misc_t *)((uintptr_t)chunk + (uintptr_t)map_misc_offset) + pageind-map_bias);
+					printf("\nelm: 0x%"PFMTx"\n", (GHT*)elm);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: dont use printf, use r_cons_printf or eprintf

File: libr/core/linux_heap_jemalloc.c [hunk 11]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+
+					arena_chunk_map_misc_t *m = R_NEW0 (arena_chunk_map_misc_t);
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: check if allocation fails

File: libr/core/linux_heap_jemalloc.c [hunk 12]
```diff
@@ -372,6 +372,89 @@ static void GH(jemalloc_get_bins)(RCore *core, const char *input) {
...
+					r_core_read_at (core, elm, (ut8*)m, sizeof(arena_chunk_map_misc_t));
```
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: missing space before (

File: libr/core/linux_heap_jemalloc.c (hunk 7 above)
Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: use PFMT64, and ut64 as much as possible. having GHT will make the code more complicated for formatting strings, etc.

Comment (radare) [inline libr/core/linux_heap_jemalloc.c]: missing spaces around +

**Comment by XVilka** [comment]
Can you please rebase this PR?

**Comment by n4x0r** [comment]
jemalloc_get_runs is yet to be completed and jemalloc_get_regions is yet to be implemented. still working on it

**Pull request merged**